### PR TITLE
Fix bug of SoftwareRender sprite and optimize it

### PR DESCRIPTION
**Pull request by daoshengmu**
@makc . I have submitted a patch to enhance the problems you mentioned at #8036. Please help me review it.

I also add some sample tests in the sandbox to provide more solid tests for software renderer. Right now, there is a problem that is related to the depth buffer between drawTriangle and drawLine. I have no idea why they can not share the same depth buffer. If you have any idea, please feel free to tell me.

Thanks.

### CODE REVIEW

File: examples/js/renderers/SoftwareRenderer.js [hunk 1]
```diff
@@ -704,68 +716,65 @@ THREE.SoftwareRenderer = function ( parameters ) {
...
 			// a
 			if( bHasNormal ) {
 				tempFace.vertexNormalsModel[ 0 ] = face.vertexNormalsModel[ 0 ];
-				tempFace.vertexNormalsModel[ 1 ] = mpN12;
-				tempFace.vertexNormalsModel[ 2 ] = mpN31;
+				tempFace.vertexNormalsModel[ 1 ] = normal1;
+				tempFace.vertexNormalsModel[ 2 ] = normal3;
 			}
-			drawTriangle( v1, mpV12, mpV31, uv1, mpUV12, mpUV31, shader, tempFace, material );
+			drawTriangle( v1, vector1, vector3, uv1, uv1, uv3, shader, tempFace, material );
```
Comment (makc) [inline examples/js/renderers/SoftwareRenderer.js]: there is something wrong here, you pass uv1 twice. same goes for the calls below.

File: examples/js/renderers/SoftwareRenderer.js [hunk 2]
```diff
@@ -704,68 +716,65 @@ THREE.SoftwareRenderer = function ( parameters ) {
...
+				var weight;
+
+				weight = (1 + v2.z) * (v2.w / v1.w) / (1 + v1.z);
+				texCoord1.copy( uv1 ).multiplyScalar( weight ).add( uv2 ).multiplyScalar( 1 / (weight + 1) );
+				weight = (1 + v3.z) * (v3.w / v2.w) / (1 + v2.z);
+				texCoord2.copy( uv2 ).multiplyScalar( weight ).add( uv3 ).multiplyScalar( 1 / (weight + 1) );
+				weight = (1 + v1.z) * (v1.w / v3.w) / (1 + v3.z);
+				texCoord3.copy( uv3 ).multiplyScalar( weight ).add( uv1 ).multiplyScalar( 1 / (weight + 1) );
+			
+				uv1 = texCoord1;
```
Comment (makc) [inline examples/js/renderers/SoftwareRenderer.js]: these mpXX were short for "middle points" XX. you should not be overwriting original uvs with these.

**Comment by daoshengmu** [comment]
@makc , Sorry. That is my mistake. We should keep the mpXX naming, and I have rolled back them. Please check it out.

File: examples/js/renderers/SoftwareRenderer.js [hunk 3]
```diff
@@ -749,7 +756,7 @@ THREE.SoftwareRenderer = function ( parameters ) {
 				tempFace.vertexNormalsModel[ 1 ] = mpN23;
 				tempFace.vertexNormalsModel[ 2 ] = mpN12;
 			}
-			drawTriangle( v2, mpV23, mpV12, uv2, mpUV23, mpUV12, shader, tempFace, material );
```
Comment (makc) [inline examples/js/renderers/SoftwareRenderer.js]: are you doing find-replace there without actually testing, LOL )

**Comment by daoshengmu** [comment]
Lol. I must be too tired so that I typo it.

**Comment by makc** [comment]
ok, I exhausted my mind-compiling skills and tested it by dropping in the actual project that I did perspective correction for. two findings:

№1 it needs ( and ) in
```
!(face instanceof THREE.RenderableSprite)
```
before it even attempts to work, and

№2 it then completely fails because, I guess, at 2+ levels of recursion it starts to overwrite these vectors that I told you we can avoid to create :shit: sorry

**Comment by makc** [comment]
I guess it can still be solved using vectors pool, like we did in flash times. This would eat the memory permanently, but at least less work for the garbage collector.

**Comment by daoshengmu** [comment]
Well. That is so weird. I understand we should add () for ```!(face instanceof THREE.RenderableSprite)```. but after adding (), the vectors need to be recreated, otherwise, it will fail. Probably, it is overwritten. Like you said.

What do you mean about vectors pool? I will spend time in finding the solution tomorrow.

**Comment by makc** [comment]
means you have a bunch of vectors ready for use, you can take one, use it and put it back, and if you run out of vectors you make more. this way memory is never released, but if you are not doing something seriously wrong, it will eventually stop growing too.

**Comment by mrdoob** [comment]
https://github.com/mrdoob/three.js/blob/master/examples/js/renderers/Projector.js#L762-L776

**Comment by daoshengmu** [comment]
@makc I use vector pool, and it looks good. Check it out!

**Comment by makc** [comment]
@daoshengmu where do you decrement back mpNPoolCount, for example? the idea is that you need to return used vectors back to the pool somehow or else it will keep growing forever

**Comment by mrdoob** [comment]
@makc looks good?

**Comment by makc** [comment]
yep, works in my project.

**Comment by mrdoob** [comment]
Thanks guys!

**Pull request merged**